### PR TITLE
SWC-3836 - Invite a new user to join a team (Phase 1)

### PR DESCRIPTION
**Pull request by hhu94**
- [x] Allow users to send email invitations to new users.
- [x] Handle invitations links sent out to new user.
  - [x] Signing in
  - [x] Registering a new account
- [x] Tests

### CODE REVIEW

**Comment by jay-hodgson** [comment]
Refactoring looks good so far.  But why do you go through the Suggestion provider when the user enters an email address?  Does this approach mean that the user needs to select the email address in the dropdown suggestion list after they paste it into the textbox?  
Alternatively, could you not look at the textbox text to see if it matches your email regex.  And if so, move forward with the invitation using an email (rather than depending on the suggestion provider callback, which now includes the email from your changes?).

File: pom.xml [hunk 1]
```diff
@@ -633,7 +633,13 @@
 		    <artifactId>jira-rest-java-client</artifactId>
 		    <version>2.0.0-m2</version>
 		</dependency>
-	</dependencies>
+		<dependency>
+			<groupId>com.googlecode.gwtquery</groupId>
```
Comment (jay-hodgson) [inline pom.xml]: How much does this add to the client download (the gwt wrapper for jquery)?  If we can incorporate (to avoid "callback hell" by using promises), maybe it's worth it.

Comment (jay-hodgson) [reply pom.xml]: Is it well maintained, does it lock us into a version of jquery that it wraps?

File: src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java [hunk 2]
```diff
@@ -243,315 +93,506 @@
...
-	
-	public PasswordResetPresenter getPasswordResetPresenter();
-	
-	public RegisterAccountPresenter getRegisterAccountPresenter();
+	BulkPresenterProxy getBulkPresenterProxy();
 
-	public ProfilePresenter getProfilePresenter();
+	GlobalApplicationState getGlobalApplicationState();
+
+	PresenterProxy<HomePresenter, Home> getHomePresenter();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java:259]: Why all the extra newlines?  Like the removal of implicit "public" in interface, but the large amount of cleanup-related changes should probably be in a separate PR.

File: src/main/java/org/sagebionetworks/web/client/PortalGinModule.java [hunk 3]
```diff
@@ -163,342 +46,48 @@
...
-import org.sagebionetworks.web.client.widget.discussion.ForumWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.ForumWidgetViewImpl;
-import org.sagebionetworks.web.client.widget.discussion.NewReplyWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.NewReplyWidgetViewImpl;
-import org.sagebionetworks.web.client.widget.discussion.ReplyWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.ReplyWidgetViewImpl;
-import org.sagebionetworks.web.client.widget.discussion.SingleDiscussionThreadWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.SingleDiscussionThreadWidgetViewImpl;
-import org.sagebionetworks.web.client.widget.discussion.SubscribersWidgetView;
-import org.sagebionetworks.web.client.widget.discussion.SubscribersWidgetViewImpl;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/PortalGinModule.java]: Again, nice cleanup, but as soon as anyone runs "Organize Imports" within the IDE all of these will show up again.

File: src/main/java/org/sagebionetworks/web/client/SynapseClient.java [hunk 4]
```diff
@@ -372,7 +375,9 @@ EntityBundlePlus getEntityBundlePlusForVersion(String entityId, Long versionNumb
 
 	Entity moveEntity(String entityId, String newParentEntityId) throws RestServiceException;
 
-	String getUserIdFromUsername(String username) throws RestServiceException;
+    Team getTeam(String teamId) throws RestServiceException;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/SynapseClient.java]: Whitespace is off for these (if these are being cleaned up, then this should also be right).

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+									AuthenticationController authController,
+									GlobalApplicationState globalApplicationState) {
+		this.view = view;
+		this.registerWidget = registerWidget;
+		this.synapseClient = synapseClient;
+		this.synapseAlert = synapseAlert;
+		this.authController = authController;
+		this.placeChanger = globalApplicationState.getPlaceChanger();
+		view.setPresenter(this);
+		registerWidget.enableEmailAddressField(false);
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: Initialization of view state should not be in the constructor (since state will be lost when it's reused).  Please move to init (in setPlace)

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 6]
```diff
@@ -1609,24 +1538,72 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
...
 
 			return returnList;
 		} catch (SynapseException e) {
 			throw ExceptionUtil.convertSynapseException(e);
 		} 
 	}
-	
+
+	@Override
+	public MembershipInvtnSubmission getMembershipInvitation(MembershipInvtnSignedToken token) throws RestServiceException {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java]: Simple new methods like this should be done in the SynapseJavascriptClient instead of introducing new GWT RPC calls.

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			@Override
+			public void f() {
+				Throwable throwable = arguments(0);
+				synapseAlert.handleException(throwable);
+			}
+		});
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		PromiseRPC<InviteeVerificationSignedToken> promise = new PromiseRPC<>();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: Please format new code, indentation should be done in either tabs and spaces (I think default for Eclipse Format or Correct Indentation uses tabs).

File: src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java [hunk 8]
```diff
@@ -138,6 +131,9 @@ public void onSuccess(String sessionToken) {
 				view.setLoading(false);
 				//success, send to login place to continue login process (sign terms of use...)
 				view.showInfo(DisplayConstants.ACCOUNT_CREATED, "");
+				if (accountCreationToken.getEncodedMembershipInvtnSignedToken() != null) {
+					globalAppState.setLastPlace(new EmailInvitation(accountCreationToken.getEncodedMembershipInvtnSignedToken()));
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java:97]: nifty

File: src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java (hunk 2 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java:259]: I undid the new lines and kept the removal of the public keywords.

File: src/main/java/org/sagebionetworks/web/client/PortalGinModule.java (hunk 3 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/PortalGinModule.java]: I fixed all my imports to match that of Eclipse's auto importing behavior.

File: src/main/java/org/sagebionetworks/web/client/SynapseClient.java (hunk 4 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/SynapseClient.java]: I fixed all my files to be indented with tabs. (Sometimes using spaces for alignment). I wish there was a way to show github diffs with tabsize = 4 instead of 8, because that's what I use in my IDE (Intellij IDEA).

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 5 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: I changed how the view is initialized, but I'd like to discuss what I did further in person.

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 7 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: This should be fixed

File: src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java (hunk 8 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java:97]: This actually took me a bit to figure out haha.

File: src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java [hunk 9]
```diff
@@ -251,240 +251,327 @@
...
+	OpenSubmissionWidget getOpenSubmissionWidget();
+	DateTimeUtils getDateTimeUtils();
+	AccessorGroupWidget getAccessorGroupWidget();
+	AccessRequirementWidget getAccessRequirementWidget();
+	SelfSignAccessRequirementWidget getSelfSignAccessRequirementWidget();
+
+	SubjectWidget getSubjectWidget();
+	TeamMemberRowWidget getTeamMemberRowWidget();
+	RequestBuilderWrapper getRequestBuilder();
+=======
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java]: Doesn't this fail to compile?  (these are usually introduced during a complex merge)

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+	private RegisterWidget registerWidget;
+	private SynapseJavascriptClient jsClient;
+	private SynapseFutureClient futureClient;
+	private SynapseAlert synapseAlert;
+	private AuthenticationController authController;
+	private PlaceChanger placeChanger;
+
+	@Inject
+	public EmailInvitationPresenter(EmailInvitationView view,
+									RegisterWidget registerWidget,
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: all this cleanup, and we get 9 tabs indentation here? :)

File: src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java (hunk 9 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/PortalGinInjector.java]: Yes, I didn't make the merge correctly, fixing it up now.

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 10 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: This looks aligned when viewed with 4 space tabs. 

Should I avoid this kind of indentation?

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+					directExecutor()
+			);
+	}
+
+	private void initializeView(final MembershipInvtnSubmission mis) {
+		view.show();
+		registerWidget.enableEmailAddressField(false);
+		view.setRegisterWidget(registerWidget.asWidget());
+		view.setSynapseAlertContainer(synapseAlert.asWidget());
+		EmailInvitationPresenter.this.registerWidget.setEncodedMembershipInvtnSignedToken(encodedMISignedToken);
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: "EmailInvitationPresenter.this." necessary?

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+						}
+					},
+					directExecutor()
+			);
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		jsClient.getInviteeVerificationSignedToken(misId)
+			.transformAsync(
+					jsClient::updateInviteeId,
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: I need your help understanding this.
"jsClient::updateInviteeId"
Could you explain this syntax?
Calling getInviteeVerificationSignedToken with the MembershipInvtnSubmission ID binds it to the currently authenticated user?

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 10 above)
Comment (jay-hodgson) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: I'd recommend having the first parameter on a newline, and just indent all of these dependencies with a single tab.  But it's trivial, so feel free to ignore :)

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 12 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: This is passing updateInviteeId as a method reference to transformAsync. It's equivalent to the lambda "token -> jsClient.updateInviteeId(token)". Since the first parameter of transformAsync (AsyncFunction) is a functional interface, it can take in a method reference with one parameter (like above), a lambda, or an instance of an AsyncFunction class.

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 13]
```diff
@@ -1610,24 +1627,62 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
...
 
 			return returnList;
 		} catch (SynapseException e) {
 			throw ExceptionUtil.convertSynapseException(e);
 		} 
 	}
-	
+
+	@Override
+	public InviteeVerificationSignedToken getInviteeVerificationSignedToken(String membershipInvitationId) throws RestServiceException {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java]: please remove if in js client

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 14]
```diff
@@ -1610,24 +1627,62 @@ public Long getOpenRequestCount(String currentUserId, String teamId)
...
+		try {
+			return synapseClient.getInviteeVerificationSignedToken(membershipInvitationId);
+		} catch (SynapseException e) {
+			throw ExceptionUtil.convertSynapseException(e);
+		}
+	}
+
+	@Override
+	public void updateInviteeId(String membershipInvitationId, InviteeVerificationSignedToken token) throws RestServiceException {
+		org.sagebionetworks.client.SynapseClient synapseClient = createSynapseClient();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java]: please remove if in js client

File: src/main/java/org/sagebionetworks/web/client/widget/team/InviteWidget.java [hunk 15]
```diff
@@ -24,13 +25,14 @@
 	private GWTWrapper gwt;
 	private SynapseAlert synAlert;
 	private SynapseSuggestBox peopleSuggestWidget;
-	
+	private static RegExp emailRegExp = RegExp.compile("[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,4}");
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/team/InviteWidget.java]: Use WebConstants.VALID_EMAIL_REGEX

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java (hunk 12 above)
Comment (jay-hodgson) [reply src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: Please change to lambda

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+						public void onFailure(Throwable throwable) {
+							synapseAlert.handleException(throwable);
+						}
+					},
+					directExecutor()
+			);
+	}
+
+	private void bindInvitationToAuthenticatedUser(final String misId) {
+		jsClient.getInviteeVerificationSignedToken(misId)
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: show loading during this process

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+						public void onFailure(Throwable throwable) {
+							synapseAlert.handleException(throwable);
+						}
+					},
+					directExecutor()
+			);
+	}
+
+	private void initializeView(final MembershipInvtnSubmission mis) {
+		view.show();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java]: rename to showNotLoggedInUI()

File: src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+	public EmailInvitationPresenter(
+			EmailInvitationView view,
+			RegisterWidget registerWidget,
+			SynapseJavascriptClient jsClient,
+			SynapseFutureClient futureClient,
+			SynapseAlert synapseAlert,
+			AuthenticationController authController,
+			GlobalApplicationState globalApplicationState) {
+		this.view = view;
+		this.registerWidget = registerWidget;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/EmailInvitationPresenter.java:55]: initialize registerWidget.  
enableEmailAddressField(false), add to view.  Add synapse alert to view.

File: src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java [hunk 19]
```diff
@@ -20,40 +21,30 @@
...
 			@Override
 			public void onClick(ClickEvent event) {
-				presenter.registerUser(emailAddressField.getValue());
-			}
-		});
-		registerBtn2.addClickHandler(new ClickHandler() {
-			@Override
-			public void onClick(ClickEvent event) {
-				presenter.registerUser(emailAddressField2.getValue());
+				NewUser newUser = new NewUser();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java]: callback to the presenter with emailAddressField value.  Presenter will get encoded membership invtn signed token, and construct NewUser object.

File: src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java [hunk 20]
```diff
@@ -20,40 +21,30 @@
...
+				newUser.setEmail(emailAddressField.getValue());
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/view/users/RegisterWidgetViewImpl.java]: Presenter should probably validate email address before attempting to register.

File: src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java [hunk 21]
```diff
@@ -138,6 +136,9 @@ public void onSuccess(String sessionToken) {
 				view.setLoading(false);
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/presenter/NewAccountPresenter.java:93]: Remove place.toToken().contains("email") handling (around line 76)

File: src/main/java/org/sagebionetworks/web/client/SynapseJavascriptFactory.java [hunk 22]
```diff
@@ -199,6 +203,10 @@ public Object newInstance(OBJECT_TYPE type, JSONObjectAdapter json) throws JSONO
 			}
 
 			return list;
+		case MembershipInvtnSubmission:
+			return new MembershipInvtnSubmission(json);
+			case InviteeVerificationSignedToken:
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/SynapseJavascriptFactory.java]: check whitespace

File: src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java [hunk 23]
```diff
@@ -70,22 +70,35 @@ public void configure(List<UserProfile> profiles, List<MembershipInvtnSubmission
...
+			if (profile != null) {
+				userBadge.configure(profile);
+				td.add(userBadge);
+				if (inviteeEmail != null) {
+					Div inviteeEmailDiv = new Div();
+					inviteeEmailDiv.add(new Text(inviteeEmail));
+					td.add(inviteeEmailDiv);
+				}
+			} else if (inviteeEmail != null) {
+				userBadge.configureWithInviteeEmail(invite.getInviteeEmail());
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java]: Consider emailBadge for this

File: src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java [hunk 24]
```diff
@@ -70,22 +70,35 @@ public void configure(List<UserProfile> profiles, List<MembershipInvtnSubmission
...
+			TableData td = new TableData();
+			td.addStyleName("padding-5");
 			final UserProfile profile = profiles.get(i);
-			UserBadge renderer = ginInjector.getUserBadgeWidget();
-			MembershipInvtnSubmission invite = invitations.get(i);
-			final String inviteId = invite.getId();
+			final MembershipInvtnSubmission invite = invitations.get(i);
+			String inviteeEmail = invite.getInviteeEmail();
+			UserBadge userBadge = ginInjector.getUserBadgeWidget();
+			if (profile != null) {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidgetViewImpl.java]: Logic should be moved to presenter.  This can be deferred for this PR.

File: src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+			 xmlns:bh="urn:import:org.gwtbootstrap3.client.ui.html"
+>
+
+	<g:HTMLPanel>
+		<b:Row>
+			<b:Column size="XS_12, MD_6" offset="MD_3">
+				<g:SimplePanel ui:field="synapseAlertContainer"/>
+			</b:Column>
+		</b:Row>
+		<div class="container margin-top-15 min-height-400">
```
Comment (jay-hodgson) [inline src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml]: Remove HTMLPanel, remove div, add class names to Container below instead (addStyleName="margin-top-15 min-height-400")

File: src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+			<b:Container fluid="true">
+				<b:Row>
+					<b:Column size="SM_3"></b:Column>
+					<b:Column size="SM_6">
+						<b:Heading ui:field="invitationTitle" size="H4" addStyleNames="margin-top-20"/>
+						<b:BlockQuote ui:field="invitationMessageWrapper" visible="false">
+							<b:Heading ui:field="invitationMessage" size="H5" addStyleNames="margin-top-10"/>
+						</b:BlockQuote>
+						<b:Row>
```
Comment (jay-hodgson) [inline src/main/resources/org/sagebionetworks/web/client/view/EmailInvitationViewImpl.ui.xml]: remove Rows

File: src/main/java/org/sagebionetworks/web/client/utils/FutureUtils.java [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+		future.set(result);
+		return FluentFuture.from(future);
+	}
+
+	/**
+	 * Returns a failed future with its exception set to a new Throwable.
+	 *
+	 * @return a failed future with its exception set to a new Throwable
+	 */
+	public static <T> FluentFuture<T> getFailedFuture() {
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/utils/FutureUtils.java:63]: I wonder if this should take the exception (optionally), then we can test using different error states.

File: src/test/java/org/sagebionetworks/web/unitclient/presenter/EmailInvitationPresenterTest.java [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+		when(mis.getMessage()).thenReturn("message");
+	}
+
+	@Test
+	public void testLoggedIn() {
+		beforeSetPlace(true);
+		when(jsClient.getInviteeVerificationSignedToken(mis.getId())).thenReturn(getDoneFuture(inviteeVerificationSignedToken));
+		when(jsClient.updateInviteeId(inviteeVerificationSignedToken)).thenReturn(getDoneFuture(null));
+		presenter.setPlace(place);
+		verify(placeChanger).goTo(any(Profile.class));
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/presenter/EmailInvitationPresenterTest.java]: verify jsClient methods are called.  could capture Profile place as well, and verify it goes to the Teams area.

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/team/InviteWidgetTest.java [hunk 29]
```diff
@@ -93,8 +93,21 @@ public void testSendToMember() {
...
 		when(mockSuggestBox.getSelectedSuggestion()).thenReturn(null);
+		when(mockSuggestBox.getText()).thenReturn("notAnEmailAddress");
 		inviteWidget.validateAndSendInvite(invitationMessage);
-		verify(mockSynAlert).showError("Please select a user to send an invite to.");
+		verify(mockSynAlert).showError("Please select a user or an email address to send an invite to.");
+	}
+
+	@Test
+	public void testSendToEmailAddress() {
+		String email = "test@example.com";
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/team/InviteWidgetTest.java]: mix case :)

File: src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java [hunk 30]
```diff
@@ -79,47 +81,67 @@ public void onFailure(Throwable caught) {
...
-		invitations = new ArrayList<MembershipInvtnSubmission>();
-		createdOnDates = new ArrayList<>();
 		currentOffset = 0;
 		getNextBatch();
-	};
+	}
 
 	@Override
 	public void getNextBatch() {
 		synAlert.clear();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java]: This interaction confuses me.  I would expect that you would need to hide the More button before the rpc.

File: src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java [hunk 31]
```diff
@@ -79,47 +81,67 @@ public void onFailure(Throwable caught) {
...
 				} else {
 					synAlert.handleException(caught);
 				} 
 			}
 		});
 	}
-	
+
+	private void addInvitations(List<OpenTeamInvitationBundle> bundles) {
+		view.hideInvitations();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java]: Does this mean that all invitations hidden while these new invitation widgets are created, and then they are all shown again (if so, why?)?

Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java]: You're right, this hiding + showing is unnecessary.

File: src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java (hunk 30 above)
Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java]: I agree that we should hide the more button before the rpc, but which interaction is confusing other than that?

Comment (hhu94) [reply src/main/java/org/sagebionetworks/web/client/widget/team/OpenUserInvitationsWidget.java]: Is it the synAlert.clear()? I'm not entirely sure why it was there before the refactor either. I think it's safe to remove it.
